### ISSUE TITLE
errWithCause only serializes Error objects

### ISSUE THREAD

**Issue by darcyrush** [pinojs/pino-std-serializers#158]
Javascript can throw anything, not just Error objects. This includes primitives like strings and numbers;

```js
try {
  throw 42; // completely valid
} catch (e) {
  throw new Error("Nested error", { cause: e });
}
```

The bundled [lib.es2022.error.d.ts TypeScript typings](https://github.com/microsoft/TypeScript/blob/main/src/lib/es2022.error.d.ts#L7-L9) indirectly hints at this;

```ts
interface Error {
    cause?: unknown;
}
```
As such, I believe this is unexpected;
```ts
const serialize = require("pino-std-serializers").errWithCause;
const serialized = serialize(new Error("Example", { cause: 42 }));
console.log(serialized.cause); // undefined
```
I would expect the cause to be populated with `42`. Likewise I would expect it to be populated if a string was thrown, or whatever else is legal to be thrown in JavaScript, including objects that don't extend the JS Error class. Instead, the cause is simply lost.

This example may seem like a contrived edge case, but I have come across libraries that throw their own "Error" objects, that do not extend from the native JS Error class,

**Comment by jsumners** [comment]
> Javascript can throw anything, not just Error objects. This includes primitives like strings and numbers;

That is true, but you should _NOT_ do that. You should only throw instances of things that derive from `Error.prototype`. This library is designed with this principle in mind.

**Comment by darcyrush** [comment]
> That is true, but you should _NOT_ do that. You should only throw instances of things that derive from `Error.prototype`.

I completely agree, but best practices do not always translate to reality. 

While we can enforce this in our own code, we have no way to enforce this in external npm libraries (or even in very old nodejs core code, potentially). Wrapping every single npm library in a try catch is not simply not feasible. The majority of errors that our projects deal with are thrown from npm libraries.

Given that JS allows this - regardless of how insane it is - i think pino serialization should account for this also.

**Comment by jsumners** [comment]
Would you like to send a Pull Request to address this issue? Remember to add unit tests.

**Comment by darcyrush** [comment]
[Here is a PR](https://github.com/pinojs/pino-std-serializers/pull/159) with some questions as comments

**Comment by ifeltsweet** [comment]
We also need this because we do something like this:

```
throw new Error("Request failed", { cause:  { request } }); 
```

Which is completely valid and useful.

See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause

**Comment by darcyrush** [comment]
> We also need this because we do something like this:
> 
> ```
> throw new Error("Request failed", { cause:  { request } }); 
> ```

This is our use case too

**Comment by voxpelli** [comment]
@ifeltsweet @darcyrush The preferable way to do that would be to make a `RequestError` or similar that has a `request` as a property

Then you can log it in a more desirable way

And strictly speaking the `request` is probably not the cause of your failure, something in your code is, the `request` is merely context.

If it's a HTTP code `400` style failure then the cause is that the request didn't pass validation or similar – the cause isn't the request itself.

If it's a HTTP code `500` style failure then the cause is that the request unexpectedly made something error in your code – then the cause is that error, not the request.

Adding the `request` to the `Error` may often be redundant as well – your pino setup in your router should annotate your logs with the request context by itself. Eg. Fastify does this when you use `req.log` and when it catches a thrown error (and a thrown error will always cause a 500 response unless the error is something like a `NotFoundError`)

**Comment by voxpelli** [comment]
For historical context:

The main reason why Error causes can be any value is because all of `throw`, `Promise.reject()`, the `catch` in a try-catch etc historically can be of any value.

So locking it down to be only `Error`:s that were allowed would have necessitated people to check whether it's an `Error` before they add it as a `cause`.

See original discussion at https://github.com/tc39/proposal-error-cause/issues/21

**Comment by leppaott** [comment]
Throwing Error instances is expected but the cause is mistakingly assumed to be also an Error:

> The value of cause can be of any type. You should not make assumptions that the error you caught has an Error as its cause, in the same way that you cannot be sure the variable bound in the catch statement is an Error either. The "Providing structured data as the error cause" example below shows a case where a non-error is deliberately provided as the cause.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause

```throw new Error("Nested error", { cause: { hello: "my cause"  });```

Should work but the cause is expected to be error-like:

>   cause?: Error, // If the original error had an error.cause, it will be serialized here

**Comment by Andreyco** [comment]
> That is true, but you should _NOT_ do that. You should only throw instances of things that derive from `Error.prototype`. This library is designed with this principle in mind.

but you can `Promise.reject(42)`, right? As I stumbled on this, my attempt to address the issue - https://github.com/pinojs/pino-std-serializers/pull/204
